### DISCUSSION TITLE
Using QOpenGLWidget frame buffer for object picking

### DISCUSSION THREAD

**Discussion by asmaloney** [mosra/magnum#644] [Q&A] [answered]
After working out a [couple of issues](https://github.com/mosra/magnum/issues/643) I have a QOpenGLWidget with a camera based on the [mouse interaction example](https://doc.magnum.graphics/magnum/examples-mouseinteraction.html).

The next thing I need to do is object picking, so I've been looking at [the example](https://doc.magnum.graphics/magnum/examples-picking.html).

The example is adding a `Renderbuffer` of uint32s to store an object ID in `ColorAttachment{ 1 }`, and then using that buffer to look up the ID when the user clicks. Makes sense.

1) Can I just add a render buffer to the QOpenGLWidget's frame buffer? Something like:
    ```cpp
    mObjectIdRenderBuffer = new Magnum::GL::Renderbuffer;
    mObjectIdRenderBuffer->setStorage( Magnum::GL::RenderbufferFormat::R32UI, inFrameBuffer.viewport().size() );
    
    inFrameBuffer.attachRenderbuffer( Magnum::GL::Framebuffer::ColorAttachment{ 1 }, *mObjectIdRenderBuffer );
    
    inFrameBuffer.mapForDraw(
        { { Magnum::Shaders::PhongGL::ObjectIdOutput, Magnum::GL::Framebuffer::ColorAttachment{ 1 } } } );
    ```
    
    where `inFrameBuffer` comes from my widget:
    ```cpp
    const auto cRange = Magnum::Range2Di{ {}, { width(), height() } };
    auto qtFramebuffer = Magnum::GL::Framebuffer::wrap( defaultFramebufferObject(), cRange );
    ```

    I did this and modified the code from `PickingExample::mouseReleaseEvent` to use the QOpenGLWidget's frame buffer. While it compiles and runs (no warnings, no errors), I don't ever get a valid object ID - only 0.

2) When the QOpenGLWidget's viewport changes size do I need to resize my object ID render buffer somehow or does that happen automatically?

3) Is there a convenient way to dump or visualize the render buffers to ensure they have correct information? I'm on macOS and I haven't been able to find an OpenGL tool that lets me examine the render state. The "official" Apple one is ancient and doesn't seem to work.

Thank you!

**Comment by mosra** [comment]
> no warnings, no errors

Do you also get no GL errors? With `GL::Renderer::error()`. Frankly I have no idea if hijacking the Qt framebuffer would work, and I fear that even if it would, it could break at any random Qt update. Or there could be platforms where Qt uses the actual default framebuffer (ID = 0), and you're just lucky it's a custom one in this particular case.

> I need to resize my object ID render buffer somehow or does that happen automatically?

You have to do that, I don't think Qt expects additional buffers added to its internal framebuffer, and on resize it could even delete the whole thing and create a new framebuffer object, with a different ID.

> Is there a convenient way to dump or visualize the render buffers to ensure they have correct information?

I think [ApiTrace](https://github.com/apitrace/apitrace) should still work on Mac? I remember using it there years ago, it's in my regular toolbox on Linux and Windows and it's been invaluable.

**Comment by asmaloney** [comment]
> Do you also get no GL errors?

There is a `QOpenGLWidget::resized` signal which is called "right after the frame buffer object has been recreated due to resizing the widget". I tried to create and add a `Magnum::GL::Renderbuffer` on this signal, but I get `GL_INVALID_OPERATION`.

From the Qt docs:
> All rendering happens into an OpenGL framebuffer object. [makeCurrent](https://doc.qt.io/qt-6/qopenglwidget.html#makeCurrent)() ensure that it is bound in the context. Keep this in mind when creating and binding additional framebuffer objects in the rendering code in [paintGL](https://doc.qt.io/qt-6/qopenglwidget.html#paintGL)(). Never re-bind the framebuffer with ID 0. Instead, call [defaultFramebufferObject](https://doc.qt.io/qt-6/qopenglwidget.html#defaultFramebufferObject)() to get the ID that should be bound.

This made it sound (to me) like I could modify the default frame buffer, but maybe I'm confusing my terminology.

So if I can't modify the frame buffer I get from Qt, any suggestions for an alternate strategy? Can I create a separate frame buffer and render to it as well, and use it for object picking? Seems kind of wasteful to render twice, but 🤷 

Something like:
```cpp
void foo::draw( Magnum::GL::Framebuffer &inQtFrameBuffer )
{
// draw to inQtFrameBuffer
   inQtFrameBuffer.clear( Magnum::GL::FramebufferClear::Color | Magnum::GL::FramebufferClear::Depth );

   mCamera->draw( mDrawables );

// switch to second framebuffer somehow?
// draw to new framebuffer (with object IDs)
   mCamera->draw( mDrawables );

// switch back to inQtFrameBuffer (may not be needed)
}
```

Or is there a way to copy the frame buffer's data so we don't render twice?

(I'm clearly taking a stab in the dark here.)

> I think ApiTrace should still work on Mac?

I looked at that and there aren't any builds for it. The devs made a comment in one of the issues that they don't use/support macOS anymore. ☹️ 

Thanks again Vladimír!

**Comment by mosra** [comment, marked as answer]
Just in case you aren't aware, `Debug{} << GL::Renderer::error()` will print the human-readable enum name so you don't have to do manual string conversions or look up numeric values. Works with most Magnum enums.

If you get an invalid operation error then it's probably not a way forward. No idea what exactly could be the cause, but I personally wouldn't bother trying further with this experiment :) The way to go would be to create your own framebuffer, with the Object ID attachment, and then ultimately [blit()](https://doc.magnum.graphics/magnum/classMagnum_1_1GL_1_1AbstractFramebuffer.html#aa4b3a826d1c95beb8921b7bbb6ccb607) the color from it to the (wrapped) Qt's framebuffer. That is still one extra operation, but better than having to draw twice.

Regarding ApiTrace, I think it's just not maintained there anymore (I suppose because Apple decided to not care about GL, and it's a hostile platform in general), but the support isn't _removed_. You could try to build it yourself: https://github.com/apitrace/apitrace/blob/master/docs/INSTALL.markdown#mac-os-x

**Comment by asmaloney** [comment]
> The way to go would be to create your own framebuffer, with the Object ID attachment, and then ultimately [blit()](https://doc.magnum.graphics/magnum/classMagnum_1_1GL_1_1AbstractFramebuffer.html#aa4b3a826d1c95beb8921b7bbb6ccb607) the color from it to the (wrapped) Qt's framebuffer.

Got it!

That makes sense - use Magnum for everything, then just blit to the Qt widget.

I now recreate the Magnum Framebuffer when the widget is resized, then do something like:
```cpp
void foo::draw( Magnum::GL::Framebuffer &inFrameBuffer )
{
   startMagnumContext();
   mFrameBuffer->clearColor( 0, 0xa5c9ea_rgbf );
   mFrameBuffer->clearColor( 1, Magnum::Vector4ui{} );
   mFrameBuffer->clearDepth( 1.0f );

   mFrameBuffer->bind();

   mCamera->draw( mDrawables );
   endMagnumContext();

   Magnum::GL::AbstractFramebuffer::blit( *mFrameBuffer, inFrameBuffer, mFrameBuffer->viewport(),
                                          Magnum::GL::FramebufferBlit::Color );
}
```

One thing that tripped me up is that QOpenGLWidget has a "device pixel ratio" of 2 (at least on my machine). So I needed to convert "pixel size" to "device size" in a bunch of places.

Thanks Vladimír!

**Comment by mosra** [comment]
One additional thing, assuming `endMagnumContext()` internally does `resetState()`, I'd put it after the blit, not before -- because the blit still makes use of the state tracker. It might work this way, but also might cause strange issues later on.

**Comment by asmaloney** [comment]
Ah yes! Thank you.

The learning continues...
